File: jambeth/jAmbeth.Service/src/de/osthus/ambeth/service/ServiceByNameProvider.java
package de.osthus.ambeth.service;

import de.osthus.ambeth.ioc.IInitializingBean;
import de.osthus.ambeth.ioc.extendable.MapExtendableContainer;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.util.ParamChecker;

public class ServiceByNameProvider implements IServiceByNameProvider, IServiceExtendable, IInitializingBean
{
	@SuppressWarnings("unused")
	@LogInstance(ServiceByNameProvider.class)
	private ILogger log;

	protected MapExtendableContainer<String, Object> serviceNameToObjectMap;

	protected IServiceByNameProvider parentServiceByNameProvider;

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		if (parentServiceByNameProvider != null)
		{
			ParamChecker.assertTrue(parentServiceByNameProvider != this, "parentServiceByNameProvider");
		}

		serviceNameToObjectMap = new MapExtendableContainer<String, Object>("serviceName", "service");
	}

	public void setParentServiceByNameProvider(IServiceByNameProvider parentServiceByNameProvider)
	{
		this.parentServiceByNameProvider = parentServiceByNameProvider;
	}

	@Override
	public void registerService(Object service, String serviceName)
	{
		serviceNameToObjectMap.register(service, serviceName);
	}

	@Override
	public void unregisterService(Object service, String serviceName)
	{
		serviceNameToObjectMap.unregister(service, serviceName);
	}

	@Override
	public Object getService(String serviceName)
	{
		Object service = serviceNameToObjectMap.getExtension(serviceName);
		if (service == null && parentServiceByNameProvider != null)
		{
			service = parentServiceByNameProvider.getService(serviceName);
		}
		return service;
	}
}

File: ambeth/Ambeth.Service/ambeth/service/ServiceByNameProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Ioc.Extendable;

namespace De.Osthus.Ambeth.Service
{
    public class ServiceByNameProvider : IServiceByNameProvider, IServiceExtendable, IInitializingBean
    {
        [LogInstance]
		public ILogger Log { private get; set; }

        protected MapExtendableContainer<String, Object> serviceNameToObjectMap;

        public virtual void AfterPropertiesSet()
        {
            serviceNameToObjectMap = new MapExtendableContainer<String, Object>("serviceName", "service");
        }

        public void registerService(Object service, String serviceName)
        {
            serviceNameToObjectMap.Register(service, serviceName);
        }
        public void unregisterService(Object service, String serviceName)
        {
            serviceNameToObjectMap.Unregister(service, serviceName);
        }
        public Object GetService(String serviceName)
        {
            return serviceNameToObjectMap.GetExtension(serviceName);
        }
    }
}
